Refactoring in Graph3d

wimrijnders opened the pull request:
Please accept this pull request containing changes to `Graph3d`.

 The work done is purely refactoring only; the functionality should be identical to previous (except for the occasional bug which I fixed along the way).

 My actual goal is to enable multiple graphs within a view.
The refactoring is, in my opinion, an essential step to attain that goal.
However, I think the refactoring as such is worthwhile to fold back on its own, because it makes the `Graph3d` code more maintainable.

 The following actions have been performed:
### 1. Refactored graph handling so that code is isolated per graph style
 This is the focus of this pull request. As mentioned, my actual goal is to display multiple graphs.

### 2. Moved common code for handling options to separate module
This is module `Settings.js`. Some specific handling of options has been left in `Graph3d.js`.

### 3. Checked all code in folder 'graph3d' for compliance with the style guidelines
 This has been done as far as I can take it.

 Some deliberate exceptions:

 - *jsDoc-like comments*, e.g. '@param'. Not sure how to multiline this.
 - The occasional *URL's* in comments.
 - *const's upper case*. This sometimes makes for pretty unreadable names, I've avoided it in some cases.
 - *Function length*: I'm not going to touch the existing large functions if I don't need to; sorry.


##  Further Notes
 I have the following reservation about my changes:
I used `Object.create()` for implementing inheritance.
This is an `ECMAScript5` construct, and it might be the case that the `vis.js` project requires further backward compatibility. Please advise me on this.

 I also added a single line to `CONTRIBUTING.md`, because of a mistake I made
 when starting with changes. Please check if you are OK with this.

yotamberk commented on lib/graph3d/Graph3d.js:
```diff
@@ -351,62 +214,72 @@ Graph3d.prototype.getNumberOfColumns = function(data) {
  */
 Graph3d.prototype._dataInitialize = function (rawData, style) {
+  if (rawData === undefined)
+    return;
```
You might consider adding:
```javascript
if (rawData && !(rawData instanceof DataSet || rawData instanceof DataView)) {
throw new Error('Array, DataSet, or DataView expected');
}
```
already at this point

yotamberk commented on lib/graph3d/Graph3d.js:
```diff
@@ -1,13 +1,19 @@
-var Filter = require('./Filter');
-var Slider = require('./Slider');
-var StepNumber = require('./StepNumber');
+var Emitter      = require('emitter-component');
```
I hope i don't sound too nitpicky but I've always preferred required libraries ordered alphabetically.
You don't have to rearrange it, but I do think it looks and reads better (similar to package.json in npm).

yotamberk commented:
Regarding your reservations:
> I used Object.create() for implementing inheritance.
This is an ECMAScript5 construct, and it might be the case that the vis.js project requires further backward compatibility. Please advise me on this.

It might not be suitable to use `Object.create()`. I agree we need more backward compatibility. We might consider holding the use of any ES code before using webpack babel-loaders.

>I also added a single line to CONTRIBUTING.md, because of a mistake I made
when starting with changes. Please check if you are OK with this.

Great! I also made that mistake at the beginning so this will reduce confusion and frustration!

wimrijnders commented:
Hi yotamberk,

Thanks for your quick response! I'll process your feedback, however nitpicky it is ;-).

I understand that you prefer that someone else with more working knowledge on the Graph3d part reviews it as well; that's perfectly fine and understandable.

As to `Object.create()`, I have a Plan B ready, namely the 'old' way I was doing inheritance. I kind of hoped that the 'new' way would be acceptable, but never mind.

Thanks again,
Wim.

wimrijnders commented:
Hi,

I added the following changes:

- Added suggestions by yotamberk
- All throw statements throw Error instances
- Used postfix 'Graph' instead of 'StyleHandler' for naming. Not only is this shorter, it's also more descriptive
- Adjusted inheritance so that `Object.create()` is not used.

All this for the Graph3d part only, mind you.

wimrijnders commented:
Hi,

I noticed the conflicts message above.  What is the correct procedure? To pull the develop branch and merge with it?

Also, my work has been ongoing; I'm still aiming for the multiple graphs. Should I create a new pull request for these changes?

yotamberk commented:
Because we are merging PRs, sometimes conflicts pop up, so the best way is to pull from the develop branch in to your branch, solve the conflicts, and push to your branch (the PR will update with your conflict resolving).

I suggest you open another PR with the multiple graphs simply branched off this branch.
Good luck =)

mojoaxel commented:
@wimrijnders Thanks, but I would prefer using a "official" code style like [jsbeautifier](/tarunc/gulp-jsbeautifier) or something similar. Everything else is just taste.
Also please try to separate fixes from style changes. At the moment your code is is realy hard to review.

wimrijnders commented:
Ok thanks. I'm pretty impressed by your responsiveness BTW.

I'll have a shot at the conflict and I'll prepare a new branch for PR (TIL new acronym).

wimrijnders commented:
> Also please try to separate fixes from style changes. At the moment your code is is realy hard to review.

@mojoaxel I don't understand what you mean by this? I scrolled back and I can't figure out what you're responding to.

mojoaxel commented:
@wimrijnders @yotamberk I see some potential for problems here. To be honest: I'm not willing to merge big cahnges in code-Style. Thsi messes up the "blame" functionality and the "contribution statistiks". As I mentioned bevor: I'm more than willing to fix the styling but this should be done in the build process by using a code-formatter. Not by hand. Functional changes should be separated in multiple pull request to make them easy testable and reviewable.

@wimrijnders I know I sound like an ass but this a big project and we should not lose committers by being unprofessional. I know git it complicated but we are willing to support you where we can.
I would making a backupü of your  "develop" branch by renaming it and start all over with small functional fixes. 
The code-styling should be handled in a separate issue and mybe introduced together with the next major release!?

wimrijnders commented:
> Changes to big to review.

Hm. Ok, so please advise me how I can slice it up then.

Basically, what I did is move the graph-style specific code if if-blocks like:

```js
 if (this.style === Graph3d.STYLE.GRID ||
    this.style === Graph3d.STYLE.SURFACE) {
    this._redrawDataGrid();
  }
  else if (this.style === Graph3d.STYLE.LINE) {
    this._redrawDataLine();
  }
  else if (this.style === Graph3d.STYLE.BAR ||
    this.style === Graph3d.STYLE.BARCOLOR ||
    this.style === Graph3d.STYLE.BARSIZE) {
    this._redrawDataBar();
  }
  else {
    // style is DOT, DOTLINE, DOTCOLOR, DOTSIZE
    this._redrawDataDot();
  }
```

... to separate style-distinct classes. The internals of the methods of these classes are the same as previous. I don't see how I can split that up in piece-meal bites.

mojoaxel commented:
> @mojoaxel I don't understand what you mean by this? I scrolled back and I can't figure out what you're responding to.

You made over 2000 coed changes in one pull request. How should we check if there is no error in there? Normally every pull request only addresses one issues at a time, changing only the lines of code necessary for this single fix. Changing the code-Style of a whole project to his ohn taste is just a "no-go".

wimrijnders commented:
>  I'm not willing to merge big cahnges in code-Style

What code-style do you mean specifically? I following the styling guides almost to the letter, but I don't think that is what you mean.

mojoaxel commented:
> What code-style do you mean specifically? I following the styling guides almost to the letter, but I don't think that is what you mean.

It looks like you fixed mainly the formating of the code (whitespaces, newline ...) to make the code look more beautiful to your taste!? That's just not recommended, to say it nicely.

mojoaxel commented:
@wimrijnders To say it in another way: You should only touch those lines of code that you really need to change to accomplish the one single improvement you are working one.

wimrijnders commented:
@mojoaxel I followed the style guidelines for the most part. I also aligned assignments where applicable for better readability. This is not good? I'm not doing this for myself but for everybody.

I'm really trying to understand your point of view; your note that this is a lot of code to review I can understand. To note, I ran all the demo's with my changes to be sure that everything still works correctly.

However, I must remark that your comment that I did this pull request to 'make the code more beautiful to my taste' misses the mark entirely. The refactoring of the graph drawing is the focus, the whitespace is a side issue.

OK, I'll stop this discussion from my side. I'll leave this pull request as is and see how I can set up further pull requests so that they are better reviewable.

wimrijnders commented:
An honest question: If I want to make the code conform to the style guide, can I implement all items in the style guide in one pull request? Or should I do it one point at a time?

wimrijnders commented:
Rather than continue this pull request, I'll start new ones. I hereby regard this case closed.

mojoaxel commented:
> An honest question: If I want to make the code conform to the style guide, can I implement all items in the style guide in one pull request? Or should I do it one point at a time?

An honest answer: In my opinion You should only touch code of other committers if you really have to. Changing the source code of a project to a unique code style is a big thing and  must be discussed on a broad base (feel free to open an issue). In my opinion it makes no sense of having an code style without enforcing it in the build process. I personally will not merge pure style changes.

wimrijnders commented:
> Changing the source code of a project to a unique code style is a big thing
I agree with this. I urgently note, however, I was following the style guide.

> In my opinion it makes no sense of having an code style without enforcing it in the build process.
I can also agree with this. However, there is code which does not conform to the style guides in Graph3d.js. I would think it is an urgent matter to ensure that it conforms.

> I personally will not merge pure style changes.
You have a style guide, you ask developers to conform to it. As far as I'm concerned, because of this non-conformance is a bug.
